File: Tutorials/Codes/Modern_Control_Course/Chapter29/Lesson1/Chapter29_Lesson1.py
# Chapter29_Lesson1.py
# Linear time-varying (LTV) state-space simulation:
#     x_dot = A(t)x + B(t)u(t)
#     y     = C(t)x + D(t)u(t)
#
# Requirements:
#     pip install numpy scipy matplotlib

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


def matrices(t: float):
    """Return A(t), B(t), C(t), D(t) for a second-order LTV oscillator."""
    omega = 1.0 + 0.30 * np.sin(0.70 * t)
    damping = 0.15 + 0.05 * np.cos(0.50 * t)

    A = np.array([
        [0.0, 1.0],
        [-(omega ** 2), -2.0 * damping * omega]
    ], dtype=float)

    B = np.array([
        [0.0],
        [1.0 + 0.20 * np.sin(0.30 * t)]
    ], dtype=float)

    C = np.array([
        [1.0 + 0.10 * np.sin(0.40 * t), 0.0]
    ], dtype=float)

    D = np.array([[0.0]], dtype=float)
    return A, B, C, D


def input_signal(t: float) -> np.ndarray:
    """Scalar input written as a length-1 vector."""
    return np.array([np.sin(1.2 * t)], dtype=float)


def rhs_state(t: float, x: np.ndarray) -> np.ndarray:
    A, B, _, _ = matrices(t)
    u = input_signal(t)
    return A @ x + B @ u


def rhs_augmented_state_transition(t: float, z: np.ndarray) -> np.ndarray:
    """
    Integrate x(t) and Phi(t,t0) together.

    z = [x; vec(Phi)] where Phi is flattened column-wise.
    Phi_dot = A(t) Phi, Phi(t0,t0) = I.
    """
    n = 2
    x = z[:n]
    Phi = z[n:].reshape((n, n), order="F")
    A, B, _, _ = matrices(t)
    u = input_signal(t)

    x_dot = A @ x + B @ u
    Phi_dot = A @ Phi

    return np.concatenate([x_dot, Phi_dot.reshape(n * n, order="F")])


def main():
    t0, tf = 0.0, 20.0
    x0 = np.array([1.0, 0.0], dtype=float)
    n = len(x0)

    # Direct state integration
    sol_x = solve_ivp(rhs_state, (t0, tf), x0, dense_output=True, rtol=1e-9, atol=1e-11)

    # Joint integration of state and state-transition matrix
    Phi0 = np.eye(n)
    z0 = np.concatenate([x0, Phi0.reshape(n * n, order="F")])
    sol_z = solve_ivp(
        rhs_augmented_state_transition,
        (t0, tf),
        z0,
        dense_output=True,
        rtol=1e-9,
        atol=1e-11,
    )

    t_grid = np.linspace(t0, tf, 600)
    x_grid = sol_z.sol(t_grid)[:n, :].T

    y_grid = []
    for t, x in zip(t_grid, x_grid):
        _, _, C, D = matrices(float(t))
        u = input_signal(float(t))
        y_grid.append(float((C @ x + D @ u)[0]))
    y_grid = np.array(y_grid)

    Phi_tf_t0 = sol_z.y[n:, -1].reshape((n, n), order="F")
    print("Phi(tf,t0) =")
    print(Phi_tf_t0)
    print("x(tf) =", sol_z.y[:n, -1])
    print("y(tf) =", y_grid[-1])

    plt.figure()
    plt.plot(t_grid, x_grid[:, 0], label="x1(t)")
    plt.plot(t_grid, x_grid[:, 1], label="x2(t)")
    plt.plot(t_grid, y_grid, label="y(t)", linestyle="--")
    plt.xlabel("time")
    plt.ylabel("response")
    plt.title("Chapter 29 Lesson 1: LTV state-space response")
    plt.grid(True)
    plt.legend()
    plt.show()


if __name__ == "__main__":
    main()

File: Tutorials/Codes/Modern_Control_Course/Chapter29/Lesson1/Chapter29_Lesson1.m
% Chapter29_Lesson1.m
% Linear time-varying (LTV) state-space simulation:
%     x_dot = A(t)x + B(t)u(t)
%     y     = C(t)x + D(t)u(t)

clear; clc; close all;

tspan = [0, 20];
x0 = [1; 0];

opts = odeset('RelTol', 1e-9, 'AbsTol', 1e-11);
[t, x] = ode45(@rhs_state, tspan, x0, opts);

y = zeros(size(t));
for k = 1:length(t)
    [~, ~, C, D] = ltv_matrices(t(k));
    u = input_signal(t(k));
    y(k) = C * x(k, :)' + D * u;
end

figure;
plot(t, x(:, 1), 'LineWidth', 1.5); hold on;
plot(t, x(:, 2), 'LineWidth', 1.5);
plot(t, y, '--', 'LineWidth', 1.5);
grid on;
xlabel('time');
ylabel('response');
title('Chapter 29 Lesson 1: LTV state-space response');
legend('x_1(t)', 'x_2(t)', 'y(t)');

% Optional Simulink guidance:
% Implement A(t), B(t), C(t), D(t) as MATLAB Function blocks, use an
% Integrator block for x_dot, and feed x and u into y = C(t)x + D(t)u.

function dx = rhs_state(t, x)
    [A, B, ~, ~] = ltv_matrices(t);
    u = input_signal(t);
    dx = A * x + B * u;
end

function [A, B, C, D] = ltv_matrices(t)
    omega = 1.0 + 0.30 * sin(0.70 * t);
    damping = 0.15 + 0.05 * cos(0.50 * t);

    A = [0, 1;
        -(omega^2), -2.0 * damping * omega];

    B = [0;
        1.0 + 0.20 * sin(0.30 * t)];

    C = [1.0 + 0.10 * sin(0.40 * t), 0];
    D = 0;
end

function u = input_signal(t)
    u = sin(1.2 * t);
end
